doc/FormattingDecisionSpaces.md

```diff
@@ -0,0 +1,93 @@
+## Formatting Decision: Number of Spaces
+
+This is a document explaining our reasoning behind the formatting decision for the default number of spaces, representing an indentation.
+This document assumes that indentation is not aligned and that indentation requires a fixed number of spaces.
+The document discusses why we chose 4 spaces instead of 2.
+
+## Analysis
+
+Our analysis is very naive and we hope it is good enough, but we are open to contributions for a more sophisticated analysis.
+We have decided to analyse the OTP, WhatsApp code base and some others, since this was too big code bases we had access to.
+
+We decided to grep for a start of a function implementation `->` followed by a newline and 2 or 4 spaces `\s`.
+
+```sh
+# count 4 spaces
+$ pcregrep -Mr --include=".*\.erl" --include=".*\.hrl" "\->\n\s\s\s\s\w" . | grep "\->" | wc -l
+# count 2 spaces
+$ pcregrep -Mr --include=".*\.erl" --include=".*\.hrl" "\->\n\s\s\w" . | grep "\->" | wc -l
+```
+
+## Results
+
+OTP:
+  - 2 spaces: 16366
+  - 4 spaces: 113481
```

Comment:
Very interesting.

Reply:
Only [very recently](https://github.com/erlang/otp/commit/5b6351fdd10c7f97d414f2e988d4a6f06d7de598) they changed that in the pretty-printer, actually.

But there are still [traces of the behavior](https://github.com/erlang/otp/blob/8ef1bc22b940c085be26024af090f1eeec71eaed/lib/syntax_tools/src/prettypr.erl#L1288-L1303).

Reply:
Nice find.
I guess we should also consider 8 spaces in the data analysis then.

Reply:
You won't find it, tho… Their algorithm was "indent some stuff with 2 spaces and some stuff with 4 spaces, but if the indentation covers more than 8 spaces, then replace those 8 spaces with a tab".
That means that if you only check indentation _after a `->`_, you'll only find 4-space indentation. To find lines indented with 2 spaces you need to find `case` statements whose expression spans multiple lines (among other similar scenarios). And to find 8-space indentation, you need 2 4-space indentations (or one 4-space and 2 2-space indentations) combined.

Reply:
OTP use 4 spaces for indentation (there are some cases where half indent is used (2 spaces). It is the emacs mode that is considered the norm. With these settings:

    (setq-default indent-tabs-mode nil)
    (setq c-basic-offset 4)

Previously if there where 8 spaces or more in a row , 8 spaces was converted to a tab, but not any more.
Old code has not been reformatted so it may contain tabs.

The half indent is used for example before `when` like this:

    foo(A) 
      when is_integer(A) ->
        A+1.

But of course you should not write like that if everything including `->` fit in the same line.

So from OTP point I would say 4 spaces is the norm for indentation (absolutely not 8 spaces).

Reply:
Phew, Thank you that saves me the work of extra analysis :)